lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
+
+    /// Diagnose any errors concerning the \c @_exported, \c @_implementationOnly,
+    /// \c @testable, or \c @_private attributes, including a
+    /// non-implementation-only import of a fragile library from a resilient one.
+    void validateOptions(NullablePtr<ModuleDecl> topLevelModule, SourceFile &SF);
+
+    /// Create an \c ImportedModuleDesc from the information in this
+    /// UnboundImport.
+    ImportedModuleDesc makeDesc(ModuleDecl *module) const {
+      return ImportedModuleDesc({ declPath, module }, options,
+                                privateImportFileName);
+    }
 
-namespace {  
-  class NameBinder {    
-  public:
+  private:
+    void validatePrivate(ModuleDecl *topLevelModule);
+    void validateImplementationOnly(ASTContext &ctx);
+    void validateTestable(ModuleDecl *topLevelModule);
+    void validateResilience(NullablePtr<ModuleDecl> topLevelModule,
+                            SourceFile &SF);
+
+    /// Diagnoses an inability to import \p modulePath in this situation and, if
+    /// \p attrs is provided and has an \p attrKind, invalidates the attribute and
+    /// offers a fix-it to remove it.
+    void diagnoseInvalidAttr(DeclAttrKind attrKind, DiagnosticEngine &diags,
+                             Diag<Identifier> diagID);
+  };
+
+  /// Represents an import whose options have been checked and module has been
+  /// loaded, but its scope (if any) has not been validated.
+  struct BoundImport {
+    UnboundImport unbound;
+    ImportedModuleDesc desc;
+    ModuleDecl *module;
+    bool needsScopeValidation = false;
+
+    BoundImport(UnboundImport unbound, ImportedModuleDesc desc,
+                ModuleDecl *module, bool needsScopeValidation);
+
+    /// Validate the scope of the import, if needed.
+    void validateScope(SourceFile &SF);
+  };
+
+  class NameBinder : public DeclVisitor<NameBinder> {
     SourceFile &SF;
-    ASTContext &Context;
+    ASTContext &ctx;
 
-    NameBinder(SourceFile &SF) : SF(SF), Context(SF.getASTContext()) {}
+    /// Imports which still need their options checked, modules loaded, and
+    /// cross-imports found.
+    SmallVector<UnboundImport, 4> unboundImports;
 
+    /// Imports which still need their scoped imports validated.
+    SmallVector<BoundImport, 16> unvalidatedImports;
+
+  public:
+    NameBinder(SourceFile &SF)
+      : SF(SF), ctx(SF.getASTContext())
+    { }
+
+    // Special behavior for these decls:
+    void visitImportDecl(ImportDecl *ID);
+    void visitPrecedenceGroupDecl(PrecedenceGroupDecl *group);
+    void visitPrefixOperatorDecl(PrefixOperatorDecl *OpDecl);
+    void visitInfixOperatorDecl(InfixOperatorDecl *OpDecl);
+    void visitPostfixOperatorDecl(PostfixOperatorDecl *OpDecl);
+
+    // Ignore other decls.
+    void visitDecl(Decl *D) {}
+
+    /// Postprocess the imports this NameBinder has bound and collect them into
+    /// a vector.
+    void finishImports(SmallVectorImpl<ImportedModuleDesc> &imports);
```

Comment:
Nitpick: "into a vector" is not necessary.

Reply:
Changed to:
```
    /// Postprocess the imports this NameBinder has bound and collect them into
    /// \p imports.
```
which makes that bit actually useful.